src/Service/EolValidator.php

```diff
@@ -0,0 +1,163 @@
+     */
+    public function validateServiceEol(): array
+    {
+        $errors = [];
+
+        foreach ($this->services as $serviceName) {
+            $service = $this->serviceFactory->create($serviceName);
+            $serviceVersion = $service->getVersion();
+
+            if ($validationResult = $this->validateService($serviceName, $serviceVersion)) {
+                $errorLevel = array_key_first($validationResult);
+                $errors[$errorLevel][] = $validationResult[$errorLevel];
+            }
+        }
+        return $errors;
+    }
+
+    /**
+     * Validates a given service and version.
+     *
+     * @param string $serviceName
+     * @param string $serviceVersion
+     * @return array
+     * @throws FileSystemException
+     */
+    public function validateService(string $serviceName, string $serviceVersion) : array
+    {
+        $serviceConfigs = $this->getServiceConfigs($serviceName);
+
+        $versionConfigs = array_filter($serviceConfigs, function ($v) use ($serviceVersion) {
+            return Semver::satisfies($serviceVersion, sprintf('%s.x', $v['version']));
+        });
+
+        if (!$versionConfigs || $versionConfigs[array_key_first($versionConfigs)]['eol'] === null) {
+            return [ValidatorInterface::LEVEL_WARNING => sprintf(
+                'Unknown or invalid EOL defined for %s %s',
+                $serviceName,
+                $serviceVersion
+            )];
+        }
+
+        $eolDate = Carbon::createFromTimestamp($versionConfigs[array_key_first($versionConfigs)]['eol']);
+
+        if (!$eolDate->isFuture()) {
+            return [ValidatorInterface::LEVEL_WARNING => sprintf(
+                '%s %s has passed EOL (%s).',
+                $serviceName,
+                $serviceVersion,
+                date_format($eolDate, 'Y-m-d')
+            )];
+        } else if ($eolDate->isFuture()
+        && $eolDate->diffInMonths(Carbon::now()) <= self::NOTIFICATION_PERIOD) {
+            return [ValidatorInterface::LEVEL_NOTICE => sprintf(
+                '%s %s is approaching EOL (%s).',
+                $serviceName,
+                $serviceVersion,
+                date_format($eolDate, 'Y-m-d')
+            )];
+        }
+
+        return [];
+    }
+
+    /**
+     * Gets the EOL configurations for the current service from eol.yaml.
+     *
+     * @param string $serviceName
+     * @return array
+     * @throws FileSystemException
+     */
+    private function getServiceConfigs(string $serviceName) : array
+    {
+        if ($this->eolConfigs === null) {
+            $this->eolConfigs = [];
+            $configsPath = $this->fileList->getServiceEolsConfig();
+            if ($this->file->isExists($configsPath)) {
+                $this->eolConfigs = Yaml::parse($this->file->fileGetContents($configsPath));
+            }
+        }
+
+        return array_key_exists($serviceName, $this->eolConfigs) ? $this->eolConfigs[$serviceName] : [];
```

Comment:
This line can be changed to more readable form:
`return $this->eolConfigs[$serviceName] ?? [];`